### PR TITLE
Implement unit tests

### PR DESCRIPTION
**Pull request by leo-stan**
## What does this PR do?

Implements #11

### CODE REVIEW

File: tests/datasets/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+*Description*
+
+Example:
+        $ python __init__.py eg_arg1 eg_arg2
+
+Inputs:
+    variable_name (type): input description
+
+"""
```
Comment (CCInc) [inline tests/datasets/__init__.py:15]: what is this file?

File: tests/helpers/helpers.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    """
+    Returns the number of trainable parameters in a model
+    Args:
+        model (LightningModule): The model object
+
+    Returns:
+        (int): The number of trainable parameters
+
+    """
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
```
Comment (CCInc) [inline tests/helpers/helpers.py:14]: maybe name this utils.py or something?

File: tests/helpers/package_available.py [hunk 3]
```diff
@@ -22,6 +21,7 @@ def _package_available(package_name: str) -> bool:
 _FAIRSCALE_AVAILABLE = not _IS_WINDOWS and _package_available("fairscale")
 
 _WANDB_AVAILABLE = _package_available("wandb")
+_OPENPOINTS_AVAILABLE = _package_available("src.models.classification.openpoints_module")
```
Comment (CCInc) [inline tests/helpers/package_available.py:24]: Once openpoints becomes a seperate module we'll want to update this

File: tests/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+*Description*
+
+Example:
+        $ python __init__.py eg_arg1 eg_arg2
+
+Inputs:
+    variable_name (type): input description
+
+"""
```
Comment (CCInc) [inline tests/models/__init__.py:15]: this file needed?

File: tests/models/test_cls_pointnet++.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import hydra
+import pytest
+
+from tests.helpers.helpers import count_trainable_params
+from tests.helpers.run_if import RunIf
+
+test_cases = [("cls_pointnet++.yaml", 2, 6, 1_462_978), ("cls_pointnet++.yaml", 4, 8, 1_463_620)]
```
Comment (CCInc) [inline tests/models/test_cls_pointnet++.py:7]: I wonder if there's a cleaner way to do this, seems like pytest might be limited

File: tests/models/test_cls_pointnet++.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+@RunIf(openpoints=True)
+@pytest.mark.parametrize(
+    "cfg_model, num_classes, num_feats, num_params", test_cases, indirect=["cfg_model"]
+)
+def test_cls_pointnetpp(cfg_model, num_classes, num_feats, num_params):
+    model = hydra.utils.instantiate(cfg_model.model, num_classes=num_classes, num_feats=num_feats)
+
+    assert count_trainable_params(model) == num_params
+    assert model.net.encoder.channel_list[0] == num_feats
+    assert model.net.prediction.head[4][0].out_features == num_classes
```
Comment (CCInc) [inline tests/models/test_cls_pointnet++.py:19]: Is this the same as `head[-1]`? If so that would be preferred

File: tests/models/test_seg_pointnet++.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+@RunIf(openpoints=True)
+@pytest.mark.parametrize(
+    "cfg_model, num_classes, num_feats, num_params", test_cases, indirect=["cfg_model"]
+)
+def test_seg_pointnetpp(cfg_model, num_classes, num_feats, num_params):
+    model = hydra.utils.instantiate(cfg_model.model, num_classes=num_classes, num_feats=num_feats)
+
+    assert count_trainable_params(model) == num_params
+    assert model.net.encoder.channel_list[0] == num_feats
+    assert model.net.head.head[2][0].out_channels == num_classes
```
Comment (CCInc) [inline tests/models/test_seg_pointnet++.py:19]: same here?

File: tests/test_configs.py [hunk 8]
```diff
@@ -1,29 +1,49 @@
...
+    _test_config(cfg_train)
 
-    HydraConfig().set_config(cfg_train)
 
-    hydra.utils.instantiate(cfg_train.datamodule)
-    hydra.utils.instantiate(cfg_train.model)
-    hydra.utils.instantiate(cfg_train.trainer)
+eval_experiments = [
+    ["model=cls_pointnet++", "data=cls_modelnet2048", "ckpt_path=."],
+    ["model=seg_pointnet++", "data=seg_s3dis1x1", "ckpt_path=."],
```
Comment (CCInc) [inline tests/test_configs.py:22]: does ckpt_path need to be set?

File: tests/test_data_transforms.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from src.transforms import Compose
+from src.transforms.augment import RandomScaleAnisotropic, RandomTranslate
+
+
+def test_random_translate():
+    translate = RandomTranslate(delta=[0.5, 0.5, 0.5])
+    assert translate.delta == [0.5, 0.5, 0.5]
```
Comment (CCInc) [inline tests/test_data_transforms.py:7]: I wonder if there's a way to test that these functions are doing what they're supposed to...

Maybe giving them a data object and checking that after running the transform the position is within what it's supposed to be?

File: tests/test_eval.py [hunk 10]
```diff
@@ -2,28 +2,46 @@
...
+test_experiments = [
+    (
+        ["experiment=cls_modelnet_pointnet++"],
+        ["model=cls_pointnet++", "data=cls_modelnet2048", "ckpt_path=."],
+        1,
+    ),
+    (
+        ["experiment=seg_s3dis1x1_pointnet++"],
+        ["model=seg_pointnet++", "data=seg_s3dis1x1", "ckpt_path=."],
+        2,
```
Comment (CCInc) [inline tests/test_eval.py:20]: is there a rational to parameterizing the max epochs for testing?

**Review by clee-ai — COMMENTED** [review]
Awesome stuff, thanks!!

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -41,47 +45,51 @@ jobs:
...
+  #      - name: List dependencies
+  #        run: |
+  #          python -m pip list
+  #
+  #      - name: Run pytest
+  #        run: |
+  #          pytest -v
 
   # upload code coverage report
   code-coverage:
```
[the remark this answers is not in the document]
Comment (leo-stan) [reply .github/workflows/test.yml:49]: Future PR :smile: #42

File: .github/workflows/test.yml [hunk 12]
```diff
@@ -41,47 +45,51 @@ jobs:
...
+  #          pip install -r requirements.txt
+  #          pip install pytest
+  #
+  #      - name: List dependencies
+  #        run: |
+  #          python -m pip list
+  #
+  #      - name: Run pytest
+  #        run: |
+  #          pytest -v
```
[the remark this answers is not in the document]
Comment (leo-stan) [reply .github/workflows/test.yml]: https://github.com/CCInc/3d-ml/actions/runs/3652667808/jobs/6171315184

[the remark this answers is not in the document]
Comment (leo-stan) [reply .github/workflows/test.yml]: Not working it seems, can't really be bothered trying to fix rn :smile:

File: src/transforms/__init__.py [hunk 13]
```diff
@@ -1,17 +1,17 @@
-from typing import Any, Optional
+from typing import Any, List, Optional
 
-from torch_geometric.transforms import BaseTransform, FixedPoints
+from torch_geometric.transforms import BaseTransform
```
[the remark this answers is not in the document]
Comment (leo-stan) [reply src/transforms/__init__.py:3]: Oh yeah sure, my code optimizer removed it sorry

File: tests/datasets/__init__.py (hunk 1 above)
Comment (leo-stan) [reply tests/datasets/__init__.py:15]: Lol autogenerated

File: tests/helpers/package_available.py (hunk 3 above)
Comment (leo-stan) [reply tests/helpers/package_available.py:24]: Yehhh it's pretty ugly but works haha

File: tests/models/test_cls_pointnet++.py (hunk 5 above)
Comment (leo-stan) [reply tests/models/test_cls_pointnet++.py:7]: Yeahhh I tried to make it so that we could easily add the new models etc to the test cases

File: tests/test_eval.py (hunk 10 above)
Comment (leo-stan) [reply tests/test_eval.py:20]: Yeah for example the seg model takes more than one epoch to start seeing an improvement in accuracy

File: tests/test_configs.py (hunk 8 above)
Comment (leo-stan) [reply tests/test_configs.py:22]: Yuh https://github.com/CCInc/3d-ml/blob/main/configs/eval.yaml

File: .github/workflows/test.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (clee-ai) [reply .github/workflows/test.yml]: Hmmm yeah fun. Let's make an issue for it and I'll look at it later if I feel like it 😄

File: tests/test_eval.py (hunk 10 above)
Comment (leo-stan) [reply tests/test_eval.py]: KK I've removed that
